From aa348606fb6dc108200657525d2605da6526b97f Mon Sep 17 00:00:00 2001
From: Eric Yang <eric.yang.wiwynn@gmail.com>
Date: Thu, 6 Mar 2025 16:48:20 +0800
Subject: [PATCH] platform-mc: Add MCTP recovery option

Allow enabling or disabling MCTP 2.0 recovery feature.
Calls `.Recover` method on timeout to handle communication recovery.

Change-Id: I298a1b397495496622254de80c67e081ec104945
Signed-off-by: Eric Yang <eric.yang.wiwynn@gmail.com>
Upstream-Status: Pending
---
 common/utils.cpp                 | 4 ++++
 meson.build                      | 2 ++
 meson.options                    | 8 ++++++++
 platform-mc/terminus_manager.cpp | 2 ++
 4 files changed, 16 insertions(+)

diff --git a/common/utils.cpp b/common/utils.cpp
index d0bd39e..0a4509a 100644
--- a/common/utils.cpp
+++ b/common/utils.cpp
@@ -588,6 +588,7 @@ int emitStateSensorEventSignal(uint8_t tid, uint16_t sensorId,
 
 void recoverMctpEndpoint(const std::string& endpointObjPath)
 {
+#ifdef MCTP_RECOVERY
     auto& bus = DBusHandler::getBus();
     try
     {
@@ -605,6 +606,9 @@ void recoverMctpEndpoint(const std::string& endpointObjPath)
             "failed to make a D-Bus call to recover MCTP Endpoint, ERROR {ERR_EXCEP}",
             "ERR_EXCEP", e);
     }
+#else
+    [[maybe_unused]] auto& _endpointObjPath = endpointObjPath;
+#endif
 }
 
 uint16_t findStateSensorId(const pldm_pdr* pdrRepo, uint8_t tid,
diff --git a/meson.build b/meson.build
index cd6f800..26da906 100644
--- a/meson.build
+++ b/meson.build
@@ -118,6 +118,8 @@ elif get_option('transport-implementation') == 'af-mctp'
     conf_data.set('PLDM_TRANSPORT_WITH_AF_MCTP', 1)
 endif
 
+conf_data.set('MCTP_RECOVERY', get_option('mctp-recovery').allowed())
+
 # Firmware update inotify option
 if get_option('fw-update-pkg-inotify').allowed()
     conf_data.set('FW_UPDATE_INOTIFY_ENABLED', 1)
diff --git a/meson.options b/meson.options
index ca8920f..4c8234f 100644
--- a/meson.options
+++ b/meson.options
@@ -125,6 +125,14 @@ option(
     value: 'enabled',
 )
 
+## MCTP Recovery Option
+option(
+    'mctp-recovery',
+    type: 'feature',
+    description: 'Recovery in request timeout. Calls `.Recover` method on timeout to handle communication recovery',
+    value: 'enabled',
+)
+
 # Firmware Update configuration parameters
 option(
     'fw-update-pkg-inotify',
diff --git a/platform-mc/terminus_manager.cpp b/platform-mc/terminus_manager.cpp
index b6ed988..35774d4 100644
--- a/platform-mc/terminus_manager.cpp
+++ b/platform-mc/terminus_manager.cpp
@@ -722,6 +722,7 @@ exec::task<int> TerminusManager::sendRecvPldmMsg(
     auto rc = co_await sendRecvPldmMsgOverMctp(eid, request, responseMsg,
                                                responseLen);
 
+#ifdef MCTP_RECOVERY
     if (rc == PLDM_ERROR_NOT_READY)
     {
         // Call Recover() to check enpoint's availability
@@ -733,6 +734,7 @@ exec::task<int> TerminusManager::sendRecvPldmMsg(
         pldm::utils::recoverMctpEndpoint(endpointObjPath);
         updateMctpEndpointAvailability(mctpInfo.value(), false);
     }
+#endif
 
     co_return rc;
 }
-- 
2.53.0-Meta

